### PR TITLE
Remove locks and switch to ConcurrentDictionary implementation.

### PR DESCRIPTION
**Pull request by diger74**
Original implementation was causing deadlocks when calling Content Delivery API in some cases.

### Prerequisites

This fixes https://github.com/umbraco/Umbraco-CMS/issues/15827

### Description

The original implementation of the property source values cache was based on Dictionary plus lock objects. It was there from long time ago, it was bulletproof and wasn't causing any issues. Until Content Delivery API was introduced.

Before Content Delivery API the cache was simply containing references to other objects since it was all in memory. Delivery API brought the option to expand reference properties into objects, that in turn can have their own properties. Colliding 2 requests for related items started creating a deadlock situation.

At first, I thought the locks were trying to defend nucache read-write scenario, however after the deeper investigation I realized nucache was implemented with ConcurrentDictionary already so there shouldn't be any need for extra locks.

Then I saw a simple Dictionary usage in Property class implementation and noticed only the use of this dictionary was requiring locks. When I swapped Dictionary to ConcurrentDictionary my original issue with deadlocks has gone.

I don't have a full historical knowledge of how this implementation was used and what additional side-effects could be (I'm new to Umbraco, only a few months). If you have a wider set of test tools you can run to identify this fix is safe that would be much appreciated.

And finally, if this fix is okay, is it possible to incorporate something similar for v12 as well, please? Our project has just gone live and we don't have short term plans (budget restriction) for major versions upgrade, however installing a minor version patch is a quick thing to do.

Appreciate your time and have a nice day,
Dmitry

### CODE REVIEW

**Comment by georgebid** [comment]
Hey @diger74, thanks a lot for your PR to fix #15928, and for raising the issue. Someone on the core collaborators team will review this soon 👍

**Comment by diger74** [comment]
Hey, is there a chance someone could look at my pull request, please?

**Comment by JasonElkin** [comment]
Hey @diger74, yes!

Thanks for this, and for the AMAZING minimal reproducible example 🙌

Let me get this reviewed and tested and I'll come back to you if anything needs tweaking!

**Comment by JasonElkin** [comment]
Thanks for this @diger74, 

I've tested it and it fixes the problem as expected! (thanks again for the k6 script etc. really very helpful).

Before I go ahead and merge, I need to have a nerdy chat with HQ about performance to make sure we're doing the best we can here as ConcurrentDictionary slows things down a bit (which is to be expected). You may have noticed that I added a couple of quick wins like swapping to GetOrAdd and Lazy initialisation - this looks to improve things a bit.

**Comment by diger74** [comment]
Hi @JasonElkin, thanks for your feedback. I'm relatively new to Umbraco therefore was a bit hesitant to add anything more than just swapping dictionary implementation. Yeah, your suggestions look good, thanks for progressing!

File: src/Umbraco.PublishedCache.NuCache/Property.cs [hunk 1]
```diff
@@ -382,26 +349,18 @@ private class CacheValue
 
     private class CacheValues : CacheValue
     {
-        private Dictionary<CompositeStringStringKey, CacheValue>? _values;
+        private readonly Lazy<ConcurrentDictionary<CompositeStringStringKey, CacheValue>> _values = new();
```
Comment (Nuklon) [inline src/Umbraco.PublishedCache.NuCache/Property.cs:352]: Lazy shouldn't be necessary here.

File: src/Umbraco.PublishedCache.NuCache/Property.cs [hunk 2]
```diff
@@ -33,7 +33,7 @@ internal class Property : PublishedPropertyBase
     private object? _interValue;
 
     // the variant source and inter values
-    private Dictionary<CompositeStringStringKey, SourceInterValue>? _sourceValues;
+    private readonly Lazy<ConcurrentDictionary<CompositeStringStringKey, SourceInterValue>> _sourceValues = new();
```
Comment (Nuklon) [inline src/Umbraco.PublishedCache.NuCache/Property.cs:36]: Just create the dictionary and avoid Lazy completely?

Comment (JasonElkin) [reply src/Umbraco.PublishedCache.NuCache/Property.cs:36]: Same as above, if creating every time is more efficient overall then happy to go with that - but keen to make sure.

Comment (kjac) [reply src/Umbraco.PublishedCache.NuCache/Property.cs:36]: Agree. Lazy has historically caused a lot of trouble with disposal and memory leakage.

Comment (kjac) [reply src/Umbraco.PublishedCache.NuCache/Property.cs:36]: Hmm. Come to think of it, the `_sourceValues` dictionary is only created for variant properties. In a LOT of cases, the property is not going to be variant, so the up-front allocation of a dictionary is an overhead compared to today.

Perhaps the `Lazy` approach is justified here.

Comment (Nuklon) [reply src/Umbraco.PublishedCache.NuCache/Property.cs:36]: Just browsing a bit through the code it seems this _sourceValues is used quite a lot (either through HasValue or GetSourceValue).

**Comment by nzdev** [comment]
Probably not ideal to use ConcurrentDictionary here. https://frugalcafe.beehiiv.com/p/silver-bullet-concurrentdictionary

**Comment by JasonElkin** [comment]
Thanks @nzdev, that's a really helpful article!

In #5827, there's a reproducible. It's a race condition that can occur when two documents reference each other via properties.

DocA -> ContentPickerProperty -> DocB
DocB -> ContentPickerProperty -> DocA

If two threads access DocA and DocB simultaneously then they deadlock because each document is waiting for the other's properties to resolve.

The fix makes logical sense. Umbraco needs to support the above scenario, so the property class shouldn't be locking - and this PR removes the locks. On top of this, as @diger74 points out, the locks are _for_ the dictionary, but they also lock methods that don't  touch the dictionary at all.

We don't want to add unnecessary memory overhead, but we need _exactly_ what ConcurrentDictionary does here. The dictionary will only be used for properties with variants, and it's likely to only contain a few entries - using the constructor that set's the default capacity, as pointed out in the article should help: `new ConcurrentDictionary(-1, 5)`.  So, we won't be adding bytes and bytes of additional memory for _every_ property and we're also not using the slow/expensive APIs.

If there's a better way of caching the `SourceInterValue`s  that doesn't add so much overhead of course we'd prefer that - any ideas?

File: src/Umbraco.PublishedCache.NuCache/Property.cs [hunk 3]
```diff
@@ -33,7 +33,7 @@ internal class Property : PublishedPropertyBase
     private object? _interValue;
 
     // the variant source and inter values
-    private readonly Lazy<ConcurrentDictionary<CompositeStringStringKey, SourceInterValue>> _sourceValues = new();
+    private readonly Lazy<ConcurrentDictionary<CompositeStringStringKey, SourceInterValue>> _sourceValues = new(() => new(-1, 5));
```
Comment (Nuklon) [inline src/Umbraco.PublishedCache.NuCache/Property.cs:36]: I think this should be set depending on the expected number of entries, what's common?

**Comment by kjac** [comment]
So. V14 has been merged to the contrib branch and the contrib branch as in turn been merged to this PR. That leaves this PR in a bit of a pickle because it essentially doesn't work against V13.3 RC anymore 😢  

As we want these changes in V13.3 RC, I have ported the changes to a new PR that targets V13.3 RC. See #16023 

I'm going to close this one. Let's keep the dialog going on #16023.

File: src/Umbraco.PublishedCache.NuCache/Property.cs (hunk 3 above)
Comment (JasonElkin) [reply src/Umbraco.PublishedCache.NuCache/Property.cs:36]: Hey @Nuklon,

This SO sums up [the choice here](https://stackoverflow.com/questions/60005428/concurrentdictionary-proper-small-initial-capacity)... 

> If you want to
>
>- optimize speed: take an initial capacity which is a prime about 30% bigger than the expected maximum dictionary size. This avoids resizing.
>- optimize the memory footprint: take a prime which is about 30% bigger than the minimum expected size.
>- a balance between speed and memory footprint: Take a number in between the two from above. But in any case, take a prime.

I'll double check the most common number, but it's _low, as it's effectively number of languages x variations so 5 seemed pretty sensible here.

**Pull request closed without merging**